stl/inc/algorithm

```diff
@@ -2364,6 +2366,149 @@ namespace ranges {
+                    _Get_unwrapped(_STD move(_Last1)), _Get_unwrapped(_STD move(_First2)), _ULast2, _Pass_fn(_Pred),
+                    _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+                return _Result.in2 == _ULast2;
+            }
+        }
+
+        // clang-format off
+        template <input_range _Rng1, input_range _Rng2, class _Pr = ranges::equal_to, class _Pj1 = identity,
+            class _Pj2 = identity>
+            requires indirectly_comparable<iterator_t<_Rng1>, iterator_t<_Rng2>, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr bool operator()(
+                _Rng1&& _Range1, _Rng2&& _Range2, _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            // clang-format on
+            if constexpr (same_as<sentinel_t<_Rng2>, unreachable_sentinel_t>) {
+                return false;
+            } else if constexpr (_Sized_or_infinite_range<_Rng1> && sized_range<_Rng2>) {
+                const range_difference_t<_Rng2> _Count2 = _RANGES distance(_Range2);
+                if constexpr (sized_range<_Rng1>) {
+                    if (_Count2 > _RANGES distance(_Range1)) {
+                        return false;
+                    }
+                }
+
+                return _RANGES _Equal_count(_Ubegin(_Range1), _Ubegin(_Range2),
+                    static_cast<range_difference_t<_Rng1>>(_Count2), _Pass_fn(_Pred), _Pass_fn(_Proj1),
+                    _Pass_fn(_Proj2));
+            } else {
+                const auto _ULast2 = _Uend(_Range2);
+                const auto _Result = _RANGES _Mismatch_4(_Ubegin(_Range1), _Uend(_Range1), _Ubegin(_Range2), _ULast2,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+                return _Result.in2 == _ULast2;
+            }
+        }
+    };
+
+    inline constexpr _Starts_with_fn starts_with{_Not_quite_object::_Construct_tag{}};
+
+    // VARIABLE ranges::ends_with
+    class _Ends_with_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, input_iterator _It2, sentinel_for<_It2> _Se2,
+            class _Pr = ranges::equal_to, class _Pj1 = identity, class _Pj2 = identity>
+            requires (forward_iterator<_It1> || sized_sentinel_for<_Se1, _It1>)
+                  && (forward_iterator<_It2> || sized_sentinel_for<_Se2, _It2>)
+                  && indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr bool operator()(_It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred = {},
+            _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            // clang-format on
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+
+            if constexpr (same_as<_Se1, unreachable_sentinel_t> || same_as<_Se2, unreachable_sentinel_t>) {
+                return false;

```

Comment:
I failed to think this through, and I'm now of the opinion that `ends_with` isn't useful when either range is infinite. Ideally we'd `static_assert` that neither sentinel type is `unreachable_sentinel_t`, but the Standard doesn't permit us to do so. I think we should just give up here, ignore the existence of potentially infinite ranges, and let the UB fall where it may.

We could maybe reasonably return `false` when the second range is sized (therefore known-finite) and the first uses `unreachable` (therefore known-non-finite) but I'm not sure there's any good reason to do so.

Reply:
Yes, it seems arbitrary for me to claim that `ranges::starts_with(views::iota(0), views::iota(0))` makes sense (and should never return) but that `ranges::ends_with(views::iota(0), views::iota(0))` is meaningless, but the fact is that the former devolves to "these infinite sequences are equal" - which we can implement easily - and the latter is "from some point in the left range, all remaining elements are equal to the elements of the right range" which we can't. It's especially important that we not pessimize the performance of "proper" calls with finite ranges trying to define meaning for infinite ranges.

Reply:
My proposal would be to write a LWG issue, that explicitly resolves those corner cases rather than us arguing about it.

My proposal would be:
* starts_with: If both ranges are infinite return false
* ends_with: If either of the ranges is infinite return false

Otherwise we could add a `Constraints:`

That would give a clear meaning to the corner cases und avoids blocking calls to this function.

Reply:
I don't think we're likely to get traction with an LWG issue trying to define behavior for infinite ranges for just these two algorithms. To be clear: passing infinite ranges to Standard Library algorithms is formally UB ([iterator.requirements.general]/10 and /12). I don't think it's a problem if implementations (like ours) choose to implement behaviors that we feel are an improvement over "just ignore it and what happens happens", but I'd be nervous about implementing special case guarantees for an algorithm or two that we might have to change when and if we do get a holistic treatment for infinite ranges in the Standard Library.